llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp

```diff
@@ -897,7 +897,20 @@ Instruction *InstCombinerImpl::visitTrunc(TruncInst &Trunc) {
     }
   }
 
-  return nullptr;
+  bool Changed = false;
+  if (!Trunc.hasNoSignedWrap() &&
+      ComputeMaxSignificantBits(Src, /*Depth=*/0, &Trunc) <= DestWidth) {
+    Trunc.setHasNoSignedWrap(true);
+    Changed = true;
+  }
+  if (!Trunc.hasNoUnsignedWrap() &&
+      MaskedValueIsZero(Src, APInt::getBitsSetFrom(SrcWidth, DestWidth),
+                        /*Depth=*/0, &Trunc)) {
+    Trunc.setHasNoUnsignedWrap(true);
+    Changed = true;
```

Comment:
Can we infer this in SimplifyDemanded instead, where we already have the KnownBits of the trunc arg?

Reply:
We cannot infer nsw flags from KnownBits (e.g., `trunc (ashr i64 X, 32) to i32`). BTW we never set poison-generating flags in `SimplifyDemanded`.

Reply:
We can't infer nsw, but we can infer nuw. Do you see any reason why doing this in SimplifyDemanded would be problematic?

Reply:
> We can't infer nsw, but we can infer nuw.

I prefer to infer both flags here, then we may reuse KnownBits in further patches.

> Do you see any reason why doing this in SimplifyDemanded would be problematic?

`SimplifyDemanded` is context-sensitive. IMO it is not the right place to infer flags.

Reply:
Ping @nikic

Reply:
Okay, I won't insist on this.